tmail-backend/blob/secondary-blob-store/src/main/java/com/linagora/tmail/blob/secondaryblobstore/SecondaryBlobStoreDAO.java

```diff
@@ -0,0 +1,116 @@
+            return secondBlobStoreDAO.read(bucketName, blobId);
+        }
+    }
+
+    @Override
+    public Mono<InputStream> readReactive(BucketName bucketName, BlobId blobId) {
+        return Mono.from(firstBlobStoreDAO.readReactive(bucketName, blobId))
+            .onErrorResume(ex -> {
+                LOGGER.error("Fail to read from the first blob store with bucket name {} and blobId {}. Use second blob store", bucketName.asString(), blobId.asString(), ex);
+                return Mono.from(secondBlobStoreDAO.readReactive(bucketName, blobId));
+            });
+    }
+
+    @Override
+    public Mono<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return Mono.from(firstBlobStoreDAO.readBytes(bucketName, blobId))
+            .onErrorResume(ex -> {
+                LOGGER.error("Fail to read from the first blob store with bucket name {} and blobId {}. Use second blob store", bucketName.asString(), blobId.asString(), ex);
+                return Mono.from(secondBlobStoreDAO.readBytes(bucketName, blobId));
+            });
+    }
+
+    @Override
+    public Mono<Void> save(BucketName bucketName, BlobId blobId, byte[] data) {
+        return Mono.from(firstBlobStoreDAO.save(bucketName, blobId, data))
+            .then(Mono.from(secondBlobStoreDAO.save(bucketName, blobId, data)));
+    }
+
+    @Override
+    public Mono<Void> save(BucketName bucketName, BlobId blobId, InputStream inputStream) {
+        return Mono.fromCallable(() -> IOUtils.toByteArray(inputStream))
+            .flatMap(data -> Mono.from(firstBlobStoreDAO.save(bucketName, blobId, data))
+                .then((Mono.from(secondBlobStoreDAO.save(bucketName, blobId, data)))));
+    }
+
+    @Override
+    public Mono<Void> save(BucketName bucketName, BlobId blobId, ByteSource content) {
+        return Mono.from(firstBlobStoreDAO.save(bucketName, blobId, content))
+            .then(Mono.from(secondBlobStoreDAO.save(bucketName, blobId, content)));
+    }
+
+    @Override
+    public Mono<Void> delete(BucketName bucketName, BlobId blobId) {
+        return Mono.from(firstBlobStoreDAO.delete(bucketName, blobId))
+            .then(Mono.from(secondBlobStoreDAO.delete(bucketName, blobId)));
+    }
+
+    @Override
+    public Mono<Void> delete(BucketName bucketName, Collection<BlobId> blobIds) {
+        return Mono.from(firstBlobStoreDAO.delete(bucketName, blobIds))
+            .then(Mono.from(secondBlobStoreDAO.delete(bucketName, blobIds)));
+    }
+
+    @Override
+    public Mono<Void> deleteBucket(BucketName bucketName) {
+        return Mono.from(firstBlobStoreDAO.deleteBucket(bucketName))
+            .then(Mono.from(secondBlobStoreDAO.deleteBucket(bucketName)));
+    }
+
+    @Override
+    public Flux<BucketName> listBuckets() {
+        return Flux.from(firstBlobStoreDAO.listBuckets())
```

Comment:
The choice to list from only primary site is likely the good one but impact **NEEDS** to be documented...

 - GC can only run when primary site is up
 - and objects not replicated on primary site aren't gced, which is likely fine!